### PR TITLE
Use "cohort of" and "quantum of" rather than array indices

### PR DESCRIPTION
**Pull request by jessealama**
Also, restructure definitions of "is zero" and "is finite" for Decimal128 values.

Partial fix for #155

### CODE REVIEW

**Comment by jessealama** [comment]
(This is just an editorial change.)

**Review by nicolo-ribaudo — APPROVED** [review]
This looks good, but as an end goal we should work towards _v_ being _alwyas_ a real number rather than sometimes a decimal and sometimes a real.

File: spec.emu [hunk 1]
```diff
@@ -143,8 +143,8 @@ location: https://github.com/tc39/proposal-decimal/
         1. If _argument_ is *NaN*<sub>𝔻</sub>, return *"NaN"*.
         1. If _argument_ is *+∞*<sub>𝔻</sub>, return *"Infinity"*.
         1. If _argument_ is *-∞*<sub>𝔻</sub>, return *"-Infinity"*.
-        1. Let _v_ be _argument_[0].
-        1. Let _q_ be _argument_[1].
+        1. Let _v_ be the cohort of _argument_.
+        1. Let _q_ be quantum of _argument_.
```
Comment (ptomato) [inline spec.emu:147]:
```suggestion
        1. Let _q_ be the quantum of _argument_.
```

**Review by ptomato — APPROVED** [review]
To do this, I think we need to have a definition of "the quantum of _d_" in the introduction. It's currently not there. We also have some inconsistency between "the cohort of _d_" and "cohort(_d_)" there.

**Comment by jessealama** [comment]
> To do this, I think we need to have a definition of "the quantum of _d_" in the introduction. It's currently not there. We also have some inconsistency between "the cohort of _d_" and "cohort(_d_)" there.

Agreed.  One difficulty though is that "cohort(d)" is understood by ecmarkup as an invocation of an AO. Currently, "cohort(d)" is just defined in prose. It could be an AO, though.

**Comment by ptomato** [comment]
You could define it the same way as e.g. floor(_x_) is defined in ecma262 (see the section "Mathematical Operations")

**Comment by jessealama** [comment]
> You could define it the same way as e.g. floor(_x_) is defined in ecma262 (see the section "Mathematical Operations")

That's a great idea. I've done that now for both cohort and quantum.

**Comment by waldemarhorwat** [comment]
> This looks good, but as an end goal we should work towards _v_ being _alwyas_ a real number rather than sometimes a decimal and sometimes a real.

In some cases, such as [an input to *RoundDecimal128*](https://github.com/tc39/proposal-decimal/issues/122#issuecomment-2160377623), then yes, _v_ should always be a real number.

In other cases, such as the definition of the «_v_, _q_» tuple, we have _v_ be a [member of the set *FiniteDecimalCohorts*](https://github.com/tc39/proposal-decimal/issues/122#issuecomment-2160377623). Here _v_ is either a nonzero real number or one of two special spec symbols which we happen to write as **+0**<sub>𝔻</sub>, **-0**<sub>𝔻</sub>.

**_v_ is never a Decimal128 value**. The 𝔻 suffix just distinguishes these symbols from **+0**<sub>𝔽</sub> and **-0**<sub>𝔽</sub>.

In my initial design (prior to the one I published as issue #122) I had _v_ always be a real number, but that caused too much confusion. Due to the need to distinguish ±0, the representation of a Decimal128 became «*s*, *v*, *q*», where the sign *s* ∈ {-1, 1}, nonnegative magnitude *v* ∈ {*n* × 10<sup>*q*</sup> | *n*, *q* ∈ ℤ, 0 ≤ *n* < 10<sup>34</sup>, -6176 ≤ *q* ≤ 6111}, and *q* is as before. One problem became the constant need to strip the sign when writing these tuples and add it back when reading them. To get the mathematical value one can't just use *v*; one needs to multiply it by *s*.

The major problem with that approach was that we need the *cohort* function throughout the spec but we can't define one that returns a real number for zero — that would coalesce +0 and -0 and produce incorrect results.

The math algorithms already must include special cases for +0 and -0, so the current definition of the «_v_, _q_» tuple is much simpler.

**Comment by ptomato** [comment]
It might prevent confusion about whether +0𝔻, -0𝔻 are Decimal values, if we were to write them as spec enums: `~positive-zero~`, `~negative-zero~`.

**Pull request merged**